src/wrappers/themis/jsthemis/test/test.js

```diff
@@ -13,173 +27,409 @@ function expect_message(message) {
+                encrypted[10] = ~encrypted[10]
+
+                assert.throws(() => secureMessageBob.decrypt(encrypted),
+                    expect_code(addon.FAIL)
+                )
+            })
+            it('handles type mismatches', function() {
+                let secureMessage = new addon.SecureMessage(privateKeyA, publicKeyB)
+                generallyInvalidArguments.forEach(function(invalid) {
+                    // null values are okay (sign/verify mode)
+                    if (invalid !== null) {
+                        assert.throws(() => new addon.SecureMessage(privateKeyA, invalid), TypeError)
+                        assert.throws(() => new addon.SecureMessage(invalid, publicKeyB), TypeError)
+                    }
+                    assert.throws(() => secureMessage.encrypt(invalid), TypeError)
+                    assert.throws(() => secureMessage.decrypt(invalid), TypeError)
+                })
+            })
+        })
+        describe('sign/verify mode', function() {
+            it('requires valid keys', function() {
+                let keyPair = new addon.KeyPair()
+                let privateKey = keyPair.private()
+                let publicKey = keyPair.public()
+                // You can omit either private or public key, but not both
+                new addon.SecureMessage(null, publicKey)
+                new addon.SecureMessage(privateKey, null)
+                assert.throws(() => new addon.SecureMessage(null, null))
+                assert.throws(() => new addon.SecureMessage(null))
+                // Empty buffers are okay too
+                new addon.SecureMessage(Buffer.from(''), publicKey)
+                new addon.SecureMessage(privateKey, Buffer.from(''))
+                // But they need to be keys
+                assert.throws(() => new addon.SecureMessage(null, Buffer.from('nope')))
+                assert.throws(() => new addon.SecureMessage(Buffer.from('not a key'), null))
+            })
+            it('signs and verifies', function() {
+                let signer = new addon.SecureMessage(privateKeyA, null)
+                let verifier = new addon.SecureMessage(null, publicKeyA)
+
+                let signed = signer.sign(testInput)
+                let verified = verifier.verify(signed)
+
+                assert.deepEqual(testInput, verified)
+            })
+            it('does not allow empty messages', function() {
+                let signer = new addon.SecureMessage(privateKeyA, null)
+                let verifier = new addon.SecureMessage(null, publicKeyA)
+
+                assert.throws(() => signer.sign(emptyArray))
+                assert.throws(() => verifier.verify(emptyArray))
+            })
+            it('leaves signed data in plaintext', function() {
+                let signer = new addon.SecureMessage(privateKeyA, null)
+                let signed = signer.sign(testInput)
+                // TODO: there has to be more idiomatic way for this check...
```

Comment:
don't they have a `substring` or similar?

Reply:
> don't they have a substring or similar?

Vanilla JS certainly does not. It's copied from WasmThemis where we have to deal with that.

However, here we know that we return Node.js' `Buffer` which should have `includes` method. So... thanks! I'll try updating this check.

Reply:
happy to help 😎